### PR TITLE
feat: Add namespace tags to FDB metrics

### PR DESCRIPTION
**Pull request by pboros**
Add namespace tags to FDB metrics. The GetInternalDatabase() call (used by CDC) will always be tagged as default_namepsace.

### CODE REVIEW

File: server/metrics/fdb.go [hunk 1]
```diff
@@ -42,31 +45,53 @@ var (
...
 )
 
-func GetFdbReqTags(reqMethodName string, tx bool) map[string]string {
+func getFdbReqTags(reqMethodName string, tx bool) map[string]string {
 	return map[string]string{
-		"method": reqMethodName,
-		"tx":     strconv.FormatBool(tx),
+		"method":    reqMethodName,
+		"tx":        strconv.FormatBool(tx),
+		"namespace": "default_namespace",
```
Comment (ovaistariq) [inline server/metrics/fdb.go:52]: Should we shorten it to just default?

Comment (ovaistariq) [reply server/metrics/fdb.go:52]: I also recommend defining it as a constant.

File: server/metrics/fdb.go [hunk 2]
```diff
@@ -42,31 +45,53 @@ var (
...
 		"method":     reqMethodName,
 		"error_code": code,
 		"tx":         strconv.FormatBool(tx),
+		"namespace":  "default_namespace",
 	}
 }
 
+func addNameSpaceToTags(ctx context.Context, tags map[string]string) map[string]string {
+	namespace, err := request.GetNamespace(ctx)
+	if err != nil {
```
Comment (ovaistariq) [inline server/metrics/fdb.go:67]: Combine the two if conditions

**Review by ovaistariq — APPROVED** [review]

File: server/request/metadata.go [hunk 3]
```diff
@@ -1,4 +1,4 @@
-package middleware
+package request
```
Comment (himank) [inline server/request/metadata.go:1]: just call the file name request.go?

File: server/services/v1/sessions.go [hunk 4]
```diff
@@ -16,6 +16,7 @@ package v1
 
 import (
 	"context"
+	"github.com/tigrisdata/tigris/server/request"
```
Comment (himank) [inline server/services/v1/sessions.go:19]: imports.

File: server/metrics/metrics.go [hunk 5]
```diff
@@ -27,6 +27,10 @@ import (
 	promreporter "github.com/uber-go/tally/prometheus"
 )
 
+const (
+	DefaultReportedNameSpace string = "default_namespace"
```
Comment (himank) [inline server/metrics/metrics.go:31]: can you move this in types.go inside the server/types package and let's call it `DefaultNamespace`

File: server/metrics/fdb.go [hunk 6]
```diff
@@ -42,31 +45,51 @@ var (
...
 		"method":     reqMethodName,
 		"error_code": code,
 		"tx":         strconv.FormatBool(tx),
+		"namespace":  DefaultReportedNameSpace,
 	}
 }
 
+func addNameSpaceToTags(ctx context.Context, tags map[string]string) map[string]string {
+	namespace, err := request.GetNamespace(ctx)
+	if err != nil && err.Error() != "Namespace not found" {
```
Comment (himank) [inline server/metrics/fdb.go:67]: can we construct a proper error when `namespace not found` and then use that for comparison here?

Comment (pboros) [reply server/metrics/fdb.go]: This is done.

File: server/metrics/metrics.go (hunk 5 above)
Comment (pboros) [reply server/metrics/metrics.go]: This is done.

File: server/services/v1/sessions.go (hunk 4 above)
Comment (pboros) [reply server/services/v1/sessions.go:19]: Done.

**Comment by pboros** [comment]
I modified this PR based on the comments:
- Added specific error for namespace not found.
- If the namespace is not known, the tag will report "unknown"
- For the admin api endpoints, no namespace tag is added
- The GetInternalDatabase is no longer instrumented
- Addressed the code review comments like file naming, imports, etc

File: server/midddleware/metrics.go [hunk 7]
```diff
@@ -33,22 +34,34 @@ import (
...
 func getErrorMethodTags(ctx context.Context, fullMethod string, methodType string, errSource string, errCode string) map[string]string {
 	metaData := metrics.GetGrpcEndPointMetadataFromFullMethod(fullMethod, methodType)
 	tags := metaData.GetSpecificErrorTags(errSource, errCode)
-	namespace, err := GetNamespace(ctx)
+	namespace, err := request.GetNamespace(ctx)
 	if err != nil {
-		ulog.E(err)
+		switch err.(type) {
+		case *request.NamespaceNotFound:
+			tags["namespace"] = "unknown"
```
Comment (himank) [inline server/midddleware/metrics.go:58]: @JigarJoshi I guess you are already adding it in auth interceptor, do we need to do it here?

File: server/metrics/fdb.go [hunk 8]
```diff
@@ -42,31 +45,53 @@ var (
...
 		"method":     reqMethodName,
 		"error_code": code,
 		"tx":         strconv.FormatBool(tx),
+		"namespace":  DefaultReportedNameSpace,
 	}
 }
 
+func addNameSpaceToTags(ctx context.Context, tags map[string]string) map[string]string {
+	namespace, err := request.GetNamespace(ctx)
+	if err != nil {
```
[the remark this answers is not in the document]
Comment (pboros) [reply server/metrics/fdb.go:67]: Done.

File: server/midddleware/metrics.go (hunk 7 above)
Comment (JigarJoshi) [reply server/midddleware/metrics.go]: this can be avoided here. auth namespace will place `unknown` if the auth fails and namespace is supplied in `""`

Comment (JigarJoshi) [reply server/midddleware/metrics.go]: in all other cases, there will always be a namespace.

File: server/metrics/fdb.go [hunk 9]
```diff
@@ -38,23 +42,42 @@ var (
...
-func GetFdbReqSpecificErrorTags(reqMethodName string, code string, tx bool) map[string]string {
+func getFdbReqSpecificErrorTags(reqMethodName string, code string, tx bool) map[string]string {
 	return map[string]string{
-		"method":     reqMethodName,
-		"error_code": code,
-		"tx":         strconv.FormatBool(tx),
+		"method":        reqMethodName,
+		"error_code":    code,
+		"tx":            strconv.FormatBool(tx),
+		"tigris_tenant": DefaultReportedTigrisTenant,
```
Comment (himank) [inline server/metrics/fdb.go:61]: do you need it here as `addTigrisTenantToTags ` is setting it?

File: server/metrics/fdb.go [hunk 10]
```diff
@@ -38,23 +42,42 @@ var (
...
 )
 
-func GetFdbReqTags(reqMethodName string, tx bool) map[string]string {
+func getFdbReqTags(reqMethodName string, tx bool) map[string]string {
 	return map[string]string{
-		"method": reqMethodName,
-		"tx":     strconv.FormatBool(tx),
+		"method":        reqMethodName,
+		"tx":            strconv.FormatBool(tx),
+		"tigris_tenant": DefaultReportedTigrisTenant,
```
Comment (himank) [inline server/metrics/fdb.go:52]: same, do we need it here?

File: server/request/request.go [hunk 11]
```diff
@@ -27,6 +26,18 @@ type NamespaceExtractor interface {
 type AccessTokenNamespaceExtractor struct {
 }
 
+type NamespaceNotFound struct {
```
[the remark this answers is not in the document]
Comment (pboros) [reply server/request/request.go]: Done.

File: server/midddleware/metrics.go [hunk 12]
```diff
@@ -33,22 +34,34 @@ import (
...
 	tags := metrics.GetPreinitializedTagsFromFullMethod(methodName, methodType)
-	namespace, err := GetNamespace(ctx)
+	namespace, err := request.GetNamespace(ctx)
 	if err != nil {
-		ulog.E(err)
+		switch err.(type) {
+		case *request.NamespaceNotFound:
+			tags["namespace"] = "unknown"
+		default:
+			ulog.E(err)
```
[the remark this answers is not in the document]
Comment (pboros) [reply server/midddleware/metrics.go]: Done.

File: server/metrics/fdb.go [hunk 13]
```diff
@@ -38,23 +42,42 @@ var (
...
 	}
 )
 
-func GetFdbReqTags(reqMethodName string, tx bool) map[string]string {
+func getFdbReqTags(reqMethodName string, tx bool) map[string]string {
 	return map[string]string{
-		"method": reqMethodName,
-		"tx":     strconv.FormatBool(tx),
+		"method":        reqMethodName,
+		"tx":            strconv.FormatBool(tx),
```
Comment (himank) [inline server/metrics/fdb.go:51]: everything is transactional, do we need this extra tag for now? Or pass it in the method for completion but no need to emit it?

File: server/metrics/requests.go [hunk 14]
```diff
@@ -33,20 +37,38 @@ func newRequestEndpointMetadata(serviceName string, methodInfo grpc.MethodInfo)
...
 func (g *RequestEndpointMetadata) GetPreInitializedTags() map[string]string {
-	return map[string]string{
-		"method":       g.methodInfo.Name,
-		"grpc_service": g.serviceName,
-		"namespace":    "default_namespace",
+	if g.serviceName == AdminServiceName {
+		return map[string]string{
+			"method":        g.methodInfo.Name,
+			"grpc_service":  g.serviceName,
+			"tigris_tenant": SystemTigrisTenantName,
```
[the remark this answers is not in the document]
Comment (pboros) [reply server/metrics/requests.go:44]: Yes.

**Review by himank — DISMISSED** [review]

**Review by himank — APPROVED** [review]

**Pull request merged**